sentry/src/main/java/io/sentry/ServerNameResolvingEventProcessor.java

```diff
@@ -0,0 +1,134 @@
+package io.sentry;
+
+import io.sentry.util.Objects;
+import java.net.InetAddress;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.FutureTask;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * An {@link EventProcessor} that sets the local ip address to an {@link
+ * SentryEvent#getServerName()} property.
+ *
+ * <p>Note - this event processor is not used by SDK by default and must be configured manually on
+ * {@link SentryOptions} if there is intention to use it.
+ */
+public final class ServerNameResolvingEventProcessor implements EventProcessor {
+
+  /**
+   * Duration of the hostname caching.
+   *
+   * @see HostnameCache
+   */
+  private static final long HOSTNAME_CACHE_DURATION = TimeUnit.HOURS.toMillis(5);
+
+  /** The global hostname cache to speed up localhost hostname resolution. */
+  private final HostnameCache hostnameCache;
+
+  public ServerNameResolvingEventProcessor() {
+    this(new HostnameCache(HOSTNAME_CACHE_DURATION));
+  }
+
+  ServerNameResolvingEventProcessor(final @NotNull HostnameCache hostnameCache) {
+    this.hostnameCache = hostnameCache;
+  }
+
+  @Override
+  public @NotNull SentryEvent process(
+      final @NotNull SentryEvent event, final @Nullable Object hint) {
+    if (event.getServerName() == null) {
+      event.setServerName(hostnameCache.getHostname());
+    }
+    return event;
+  }
+
+  /**
+   * Time sensitive cache in charge of keeping track of the hostname. The {@code
+   * InetAddress.getLocalHost().getCanonicalHostName()} call can be quite expensive and could be
+   * called for the creation of each {@link SentryEvent}. This system will prevent unnecessary costs
+   * by keeping track of the hostname for a period defined during the construction. For performance
+   * purposes, the operation of retrieving the hostname will automatically fail after a period of
+   * time defined by {@link #GET_HOSTNAME_TIMEOUT} without result.
+   */
+  static final class HostnameCache {
+    /** Time before the get hostname operation times out (in ms). */
+    private static final long GET_HOSTNAME_TIMEOUT = TimeUnit.SECONDS.toMillis(1);
+    /** Time for which the cache is kept. */
+    private final long cacheDuration;
+    /** Current value for hostname (might change over time). */
+    @Nullable private volatile String hostname;
+    /** Time at which the cache should expire. */
+    private volatile long expirationTimestamp;
+    /** Whether a cache update thread is currently running or not. */
+    private final @NotNull AtomicBoolean updateRunning = new AtomicBoolean(false);
+
+    private final @NotNull Callable<InetAddress> getLocalhost;
+
+    private HostnameCache(long cacheDuration) {
+      this(cacheDuration, InetAddress::getLocalHost);
```

Comment:
method references on Android doesn't work on all the versions and linter also does not complain :(

Reply:
https://github.com/getsentry/sentry-java/blob/main/sentry-android-core/src/main/java/io/sentry/android/core/DefaultAndroidEventProcessor.java#L103-L104